### ISSUE TITLE
How to save checkpoints in XLA

### ISSUE THREAD

**Issue by mfatih7** [pytorch/xla#4482]
Hello

I have training scripts running on CPUs and GPUs without error.
I am trying to make the scripts compatible with TPUs.

I was using the following lines to save checkpoints

```
torch.save(checkpoint, path_checkpoints_file )
```

and following lines to load checkpoints

```
checkpoint = torch.load(path_checkpoints_file, map_location=torch.device('cpu'))
    
lastEpoch = checkpoint['lastEpoch']
activeChunk = checkpoint['activeChunk']
chunk_count = checkpoint['chunk_count']

model.load_state_dict(checkpoint['model_state_dict'])    
model.to(device)

optimizer.load_state_dict(checkpoint['optimizer_state_dict'])
lr_scheduler.load_state_dict(checkpoint['lr_scheduler_state_dict'])
```

For TPUs I replaced the saving operation with

```
xm.save(checkpoint, path_checkpoints_file )
```

and the loading part with

```
checkpoint = xser.load( path_checkpoints_file )

lastEpoch = checkpoint['lastEpoch']
activeChunk = checkpoint['activeChunk']
chunk_count = checkpoint['chunk_count']

model.load_state_dict(checkpoint['model_state_dict'])    
model.to(device)

optimizer.load_state_dict(checkpoint['optimizer_state_dict'])
lr_scheduler.load_state_dict(checkpoint['lr_scheduler_state_dict'])
```

But during training, the loss remains almost constant.
Do we have a template to save and load checkpoints having models, optimizers and learning schedulers?

best regards

**Comment by JackCaoG** [comment]
does https://github.com/pytorch/xla/blob/master/API_GUIDE.md#saving-and-loading-xla-tensors help?

**Comment by mfatih7** [comment]
Thank you for the answer

The link tells how to save/load models individually.
No information is given for optimizers and learning rates.

Instead of loading the optimizer and learning rate scheduler from the checkpoint, if I redefine the optimizer and learning scheduler after loading the model, learning occurs and loss decreases normally.

Any idea?

**Comment by mfatih7** [comment]
For example, without TPUs, we save checkpoints having optimizers like

```
# Additional information
EPOCH = 5
PATH = "model.pt"
LOSS = 0.4

torch.save({
            'epoch': EPOCH,
            'model_state_dict': net.state_dict(),
            'optimizer_state_dict': optimizer.state_dict(),
            'loss': LOSS,
            }, PATH)
```

load checkpoints having optimizers like

```
model = Net()
optimizer = optim.SGD(net.parameters(), lr=0.001, momentum=0.9)

checkpoint = torch.load(PATH)
model.load_state_dict(checkpoint['model_state_dict'])
optimizer.load_state_dict(checkpoint['optimizer_state_dict'])
epoch = checkpoint['epoch']
loss = checkpoint['loss']

model.eval()
# - or -
model.train()
```

**Comment by JackCaoG** [comment]
@Liyang90 Do you have an example handy?

**Comment by Liyang90** [comment]
I don't have an existing example, but I can take a look into this.
So from the description above, the key issue seems to be that after loading the checkpoint, the connection between the optimizer and the model parameters seems to be broken, so optimizer doesn't update the model and the loss stay the same.

**Comment by mfatih7** [comment]
Thank you very much for your quick replies

Yes, you understand correctly.
Actually, I think I found the solution.

The format of my training scripts was like:

If there is no checkpoint does exist in the training directory scripts generate model, optimizer, and learning rate scheduler and save these as a checkpoint.
In the first epoch, the scripts load this initial checkpoint and after the epoch finishes it updates the checkpoint.
And for each epoch, the scripts load checkpoint at the start of epochs and updates the checkpoint at the end of epochs.

Therefore the initial checkpoint is recorded before it is loaded into the device.
All other checkpoints are saved after the variables in the checkpoint are loaded into the device. 

This was not causing to a problem if the device was CPU or GPU.
But now I see that if the device is TPU before saving the initial checkpoint the model must be transferred to TPU.
Therefore if the saving part is updated like the following the loss changes as expected and training occurs.

```
model = get_model(model_type, model_no, context_size, params.INPUT_FLOAT_TYPE, device, n)
model.to(device)
optimizer=torch.optim.Adam(model.parameters(), lr=learning_rate)
lr_scheduler = torch.optim.lr_scheduler.ReduceLROnPlateau(
                                                               optimizer,
                                                               mode='min',
                                                               factor=params.LR_SCHEDULER_FACTOR,
                                                               patience=math.ceil(patience_chunk_count),
                                                               threshold=0.0001,
                                                               threshold_mode='rel',
                                                               cooldown=0,
                                                               min_lr=0,
                                                               eps=1e-08,
                                                               verbose=True,
                                                          )         
checkpoint = {

                          'model_state_dict': model.state_dict(),
                          'optimizer_state_dict': optimizer.state_dict(),
                          'lr_scheduler_state_dict': lr_scheduler.state_dict(),
                      }
xm.save(checkpoint, path_checkpoints_file )
```

**Comment by Liyang90** [comment]
Hi @mfatih7,

Great to hear that you found a solution!

I'm still curious though, because `xm.save()` would move tensors to CPU anyway, so the model placement when saving should not matter. However, when the model is loading the loaded `state_dict`, has it been moved to TPU?

**Comment by mfatih7** [comment]
Thank you for the answer @Liyang90 

Actually, the training is very slow.
I share your doubts.
I am not sure if the training occurs in CPU or TPU.
But I did not finish the update of my scripts for TPU.
I mean there are .item() calls for example.

For GPU, it is easy to observe the GPU utilization in COLAB.
But for TPU, only RAM utilization and disk utilization can be observed visually in COLAB.

Should I check TPU utilization with commands?

**Comment by Liyang90** [comment]
Yes, `.item()` would slow down the training a lot. 

https://github.com/pytorch/xla/blob/master/TROUBLESHOOTING.md has useful info about some caveats using torch_xla.

I also maybe kinda reproduces your issue in our example script in `xla/test/test_train_mp_mnist.py`.

I added below to test the checkpoint saving and loading:

```
def save_and_load(model, optimizer, flags):
  state_dict = (model.state_dict(), optimizer.state_dict())
  xm.save(state_dict, "checkpoint.pt")

  loaded_state_dict = xser.load("checkpoint.pt")
  lr = flags.lr * xm.xrt_world_size()
  device = xm.xla_device()
  model = MNIST()
  model = model.to(device)
  optimizer = optim.SGD(model.parameters(), lr=lr, momentum=flags.momentum)
  # if I do `model = model.to(device)` here, the training would not update the model
  model.load_state_dict(loaded_state_dict[0])
  optimizer.load_state_dict(loaded_state_dict[1])

  return model, optimizer
  ```

It works correctly as expected. I tested keeping the model and optimizer on CPU **before** `xm.save()`, and it still works fine.

However if I make the change in the comment, it would fail like in your case.

**Comment by mfatih7** [comment]
@Liyang90

Transferring the model to TPU after the generation of the optimizer is also discussed here
https://github.com/pytorch/xla/issues/3675
It is consistent with the second part of your trial.

On the other hand, my problem is a little bit different than the first part of your trial.
In my case, the checkpoint includes model, optimizer and learning rate scheduler.
You do not save learning scheduler in your case.

I also retried saving model before transferring it to TPU and load it back.
The training still does not occur.

**Comment by mfatih7** [comment]
Hello again

In the debugging article https://cloud.google.com/blog/topics/developers-practitioners/pytorchxla-performance-debugging-tpu-vm-part-1

the following script is used to get log data from host 

```
export XRT_TPU_CONFIG="localservice;0;localhost:51011"
export USE_TORCH=ON

python3 mmf_cli/run.py \
config=./projects/unit/configs/all_8_datasets/shared_dec_without_task_embedding.yaml \
dataset=glue_qnli \
model=unit \
training.batch_size=8 \
training.device=xla \
distributed.world_size=1 \
training.log_interval=100 \
training.max_updates=1500
```

I do not use logging module in my scripts.
How can I get log data for my own training script from the Host while using TPU?

`!python runTrain_n_to_n_TPU_single.py
`

How should I modify the above code that I use to start my training scripts?

**Comment by Liyang90** [comment]
Hi @mfatih7 ,

What kind of logging are you thinking of? And how are you connecting to the TPU cluster?

If the issue is quite different than the topic of this issue, would you mind closing this one and opening a separate issue?

Thanks,
Liyang

**Comment by mfatih7** [comment]
Ok I will open another issue.
This one might continue.
You can delete your message here.

**Comment by mfatih7** [comment]
@Liyang90

[Here](https://github.com/pytorch/xla/issues/4509#issuecomment-1404160996) I checked TPU type and memory properties.
Regarding to the checkpoints, this shows that model is successfully transferred to TPU and learning occurs in it.
What do you think?

**Comment by mfatih7** [comment]
Hello again

So while using single-core TPU I used `model.to(device)` to suppress the errors regarding to checkpointing.

In case of multiple TPU cores, [it is written](https://colab.research.google.com/github/pytorch/xla/blob/master/contrib/colab/multi-core-alexnet-fashion-mnist.ipynb#scrollTo=eSiwkKVLW9zO) that we must not use 

`device = xm.xla_device()` and `xmp.spawn(simple_map_fn, args=(flags,), nprocs=8, start_method='fork')` simultaneously.

So how can I save the initial checkpoint for the multi-core TPU case? Do you know if we have any examples?

[Here](https://colab.research.google.com/github/pytorch/xla/blob/master/contrib/colab/multi-core-alexnet-fashion-mnist.ipynb#scrollTo=pqx-VgEizPiF), it is also written that with `if not xm.is_master_ordinal():` and `if xm.is_master_ordinal():`
it is possible to select only the master core for operations while the others wait on the `rendezvous`.

After each epoch should I save only the models, optimizer, and lr_scheduler on the master core as a checkpoint?
